README
/**********************************************************************************************/
/*  This program is part of the Barcelona OpenMP Tasks Suite                                  */
/*  Copyright (C) 2009 Barcelona Supercomputing Center - Centro Nacional de Supercomputacion  */
/*  Copyright (C) 2009 Universitat Politecnica de Catalunya                                   */
/*                                                                                            */
/*  This program is free software; you can redistribute it and/or modify                      */
/*  it under the terms of the GNU General Public License as published by                      */
/*  the Free Software Foundation; either version 2 of the License, or                         */
/*  (at your option) any later version.                                                       */
/*                                                                                            */
/*  This program is distributed in the hope that it will be useful,                           */
/*  but WITHOUT ANY WARRANTY; without even the implied warranty of                            */
/*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the                             */
/*  GNU General Public License for more details.                                              */
/*                                                                                            */
/*  You should have received a copy of the GNU General Public License                         */
/*  along with this program; if not, write to the Free Software                               */
/*  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA            */
/**********************************************************************************************/

SUITE DESCRIPTION

The objective of the suite is to provide a collection of applications that allow to test
OpenMP tasking implementations. Most of the kernels come from existing ones from other
projects. Each of them comes with different implementations that allow to test different
possibilities of the OpenMP task model (task tiedness, cut-offs, single/multiple generators,
etc). It currently comes with the following kernels:

   + Alignment: Aligns sequences of proteins.
   + FFT: Computes a Fast Fourier Transformation.
   + Floorplan: Computes the optimal placement of cells in a floorplan.
   + Health: Simulates a country health system.
   + NQueens: Finds solutions of the N Queens problem.
   + Sort: Uses a mixture of sorting algorithms to sort a vector.
   + SparseLU: Computes the LU factorization of a sparse matrix.
   + Strassen: Computes a matrix multiply with Strassen's method.

FILE DESCRIPTION

The package directory is organized as follows:

   + ./bin: target directory when compiling the benchmarks (generated by `./configure').
   + ./common: source code for common objects.
   + ./config: config directory, assist the compilation phase.
   + ./inputs: input files for some of the benchmarks (when needed).
   + ./omp-tasks: source code for the OpenMP Task versions.
   + ./run: scripts for running the benchmarks.
   + ./serial: source code for the serial versions.
   + ./templates: template directory.

SUITE COMPILATION

Briefly, the shell commands `./configure' and  `make' should configure and build this package.

The `./configure' shell script attempts to guess correct values for various system-dependent
variables used during compilation (compiler and OpenMP specific flags). The shell script
examines the compilers installed in the system and let the user choose among them.  It uses
those values to create a `make.config' in `./config' directory which will be used during
the suite compilation.

If you need to change compilers, linkers or compilation options you can change manually the
file `./config/make.config'.

OTHER ISSUES TO TAKE INTO ACCOUNT

Some benchmarks are coded following a recursive model which makes an intensive use of stack
frames. It is usually needed to increase default stack sizes when working with high recursion
levels. Main stack size can be increased using `ulimit -s' linux command. Thread stacks are
usually controlled through runtime specific environment variable (e.g. Intel C Compiler uses
KMP_STACKSIZE=`value' for this purpose).

Developer activities and documentation are centralized in https://pm.bsc.es/projects/bots
 
Mail suggestions and bug reports to pm-tools@bsc.es, or using our ticket system in the web
site: http://pm.bsc.es/projects/bots
 


CHANGELOG.md
# BOTS ChangeLog

## _BOTS current version_ (unreleased)

### The following new features/information were added:

 * Strassen's OmpSs initial version (#147)
 * UTS's OmpSs initial version (#148)
 * N-Queens's OmpSs initial version (#144)
 * Health's OmpSs initial version (#142)
 * Multisort's OmpSs initial version (#145)
 * Floorplan's OmpSs initial version (#141)
 * Fibonacci's OmpSs initial version (#140)
 * OmpSs support (enabling to compile OmpSs applications) (#137)

### In addition, the following items were fixed and/or added:

 * Distribution files: CHANGELOG, README fixed (as markdown)
 * Script configure syntax error when using --help
 * UTS compilation problem, no 'omp.h' header
 * Common 'final' version compilation problem

## _BOTS version 1.1.2_ (released 2011-11-23)

### The following new features/information were added:

 * Choose output format option from run scripts (#132)
 * New Health's large and test inputs (#123)
 * Tied version for UTS benchmark (#131)

### In addition, the following items were fixed and/or added:

 * Health results in unsuccessful when compiled with -m 32 (#124)
 * UTS benchmark seg faults with medium data set (#128)
 * Fibonacci result overflows when input is bigger than 47 (#130)
 * Fibonacci was always returning successful (also when failing) (#129)
 * Wrong use of arrays in alignment benchmark (#126 an #127)
 * String overflows problem in common files (#122)

## _BOTS version 1.1.1_ (released 2011-2-11)

### The following new features/information were added:

 * Several benchmarks using the final OpenMP 3.1 final clause (#96)
 * Alignment benchmark has been adapted to clustalw implementation (#119)
 * get_architecture() has been implemented in all Linux-like systems (#44)
 * configure.sh script allows to specify compiler in the command line (#90)
 * A new version with single has been implemented for alignment benchmark (#63)
 * Verbosity in "run-health.sh -check" has been reduced (#112)

### In addition, the following items were fixed and/or added:

 * Script run-nqueens.sh no-shell bug fixed (#115)
 * sparselu: divison by zero in checkmat() fixed (#118)
 * CFLAGS not big enough bug has been fixed (#117)
 * Benchmark omp-tasks/nqueens.c out-of-bounds access bug has been fixed (#116)
 * UTS verification bug fixed (#113)
 * Script run-all.sh is not running nonexistent knapsack test anymore (#111)
 * Sort benchmark when seqmerge() is running for just one element bug fixed(#65)
 * All warnings has been removed from compilation (#114)

## _BOTS version 1.1_ (released 2010-5-17)

### The following new features/information were added:

 * A new benchmark has been added: Unbalanceed Tree Search (UTS) (#3, #98, #99, #102, #103, #105 & #106)
 * A README file has been added (#55)
 * Environment variables are allowed through running scripts (#109)
 * Task/Sec measurement variable has been changed by Nodes/Sec (#97)
 * Final clause versions have been added (#94)
 * Make incorpores a new wipe option which clean binnaries (#93)
 * Bots configure script has been improved (#92)
 * Improving Alignment parallel code, moving task generation inside 'if' stament (#62)
 * Additional argument specification in suite files (#57)
 * An abrigded list format has been included (#42)
 * Fibonacci uses pre-computed values for testing purposes (#24)

### In addition, the following items were fixed and/or added:

 * Verbose modes have been reviewed and synchronized with configure '--debug' option (#79, #104, #107 & #108)
 * Common compiler warning have been solved (using icc #100 and xlc #101 compilers)
 * Program name was not appearing in AIX machines (#76)
 * Package sources permissions has been changed (#71)
 * Input file reading sanity checking improvement (#91)
 * Run-script cutoff value int tied versions has been fixed (#86)
 * Message error when running outside a cpuset has been solved (#85)
 * Using a relative input file name in run-scripts (#83)
 * Support to associate specific parameters to different classes (#61)
 * Serial version only run once with multiple classes/inputs (#58)
 * Alignment non valid input file error has been solved (#82)
 * FFT checking problems has been solved (#66)
 * Fibonacci pre-computed values error has been solved (#95)
 * Fibonacci self-verification showed N/A when unsuccessful (#77)
 * Floorplan task count has been fixed (#87)
 * Cropped area in Floorplan benchmark has been fixed (#67)
 * NQueens serial execution returned a wrong result (#89)
 * Removing NQueens unneeded taskwaits (#88)
 * Avoiding to allocate temporary version in NQueens serial path (#88)
 * Sort algorihtm has been parametrized (#81)
 * Sort was not scrambling vector before ordering it (#64)
 * Sort constants have been changed (#70)
 * SparseLU checking were too strict and has been relaxed (#59)

## _BOTS version 1.0_ (released 2009-5-29)


ompss/fib/fib.c
/**********************************************************************************************/
/*  This program is part of the Barcelona OpenMP Tasks Suite                                  */
/*  Copyright (C) 2009 Barcelona Supercomputing Center - Centro Nacional de Supercomputacion  */
/*  Copyright (C) 2009 Universitat Politecnica de Catalunya                                   */
/*                                                                                            */
/*  This program is free software; you can redistribute it and/or modify                      */
/*  it under the terms of the GNU General Public License as published by                      */
/*  the Free Software Foundation; either version 2 of the License, or                         */
/*  (at your option) any later version.                                                       */
/*                                                                                            */
/*  This program is distributed in the hope that it will be useful,                           */
/*  but WITHOUT ANY WARRANTY; without even the implied warranty of                            */
/*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the                             */
/*  GNU General Public License for more details.                                              */
/*                                                                                            */
/*  You should have received a copy of the GNU General Public License                         */
/*  along with this program; if not, write to the Free Software                               */
/*  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA            */
/**********************************************************************************************/

#include "bots.h"
#include "fib.h"

#define FIB_RESULTS_PRE 41
long long fib_results[FIB_RESULTS_PRE] = {0,1,1,2,3,5,8,13,21,34,55,89,144,233,377,610,987,1597,2584,4181,6765,10946,17711,28657,46368,75025,121393,196418,317811,514229,832040,1346269,2178309,3524578,5702887,9227465,14930352,24157817,39088169,63245986,102334155};

long long fib_seq (int n)
{
	int x, y;
	if (n < 2) return n;

	x = fib_seq(n - 1);
	y = fib_seq(n - 2);

	return x + y;
}

#if defined(IF_CUTOFF)

long long fib (int n,int d)
{
	long long x, y;
	if (n < 2) return n;

	#pragma omp task untied shared(x) firstprivate(n) if(d < bots_cutoff_value)
	x = fib(n - 1,d+1);

	#pragma omp task untied shared(y) firstprivate(n) if(d < bots_cutoff_value)
	y = fib(n - 2,d+1);

	#pragma omp taskwait
	return x + y;
}

#elif defined(FINAL_CUTOFF)

long long fib (int n,int d)
{
	long long x, y;
	if (n < 2) return n;

	#pragma omp task untied shared(x) firstprivate(n) final(d+1 >= bots_cutoff_value) mergeable
	x = fib(n - 1,d+1);

	#pragma omp task untied shared(y) firstprivate(n) final(d+1 >= bots_cutoff_value) mergeable
	y = fib(n - 2,d+1);

	#pragma omp taskwait
	return x + y;
}

#elif defined(MANUAL_CUTOFF)

long long fib (int n, int d)
{
	long long x, y;
	if (n < 2) return n;

	if ( d < bots_cutoff_value ) {
		#pragma omp task untied shared(x) firstprivate(n)
		x = fib(n - 1,d+1);

		#pragma omp task untied shared(y) firstprivate(n)
		y = fib(n - 2,d+1);

		#pragma omp taskwait
	} else {
		x = fib_seq(n-1);
		y = fib_seq(n-2);
	}

	return x + y;
}

#else

long long fib (int n)
{
	long long x, y;
	if (n < 2) return n;

	#pragma omp task untied shared(x) firstprivate(n)
	x = fib(n - 1);
	#pragma omp task untied shared(y) firstprivate(n)
	y = fib(n - 2);

	#pragma omp taskwait
	return x + y;
}

#endif

static long long par_res, seq_res;

void fib0 (int n)
{
#if defined(MANUAL_CUTOFF) || defined(IF_CUTOFF) || defined(FINAL_CUTOFF)
	par_res = fib(n,0);
#else
	par_res = fib(n);
#endif
	bots_message("Fibonacci result for %d is %lld\n",n,par_res);
}

void fib0_seq (int n)
{
	seq_res = fib_seq(n);
	bots_message("Fibonacci result for %d is %lld\n",n,seq_res);
}

long long fib_verify_value(int n)
{
	if (n < FIB_RESULTS_PRE) return fib_results[n];
	return ( fib_verify_value(n-1) + fib_verify_value(n-2));
}

int fib_verify (int n)
{
	int result;

	if (bots_sequential_flag)
	{
		if (par_res == seq_res) result = BOTS_RESULT_SUCCESSFUL;
		else result = BOTS_RESULT_UNSUCCESSFUL;
	}
	else
	{
		seq_res = fib_verify_value(n);
		if (par_res == seq_res) result = BOTS_RESULT_SUCCESSFUL;
		else result = BOTS_RESULT_UNSUCCESSFUL;
	}

	return result;
}

